docs/developer-docs/security/security-best-practices/inter-canister-calls.mdx

```diff
@@ -269,11 +279,71 @@ mod test {
 }
 ```
 
+</TabItem>
+
+<TabItem value="motoko" label="Motoko">
+
+```motoko
+import Result "mo:base/Result";
+import Map "mo:base/HashMap";
+import Error "mo:base/Error";
+import Principal "mo:base/Principal";
+
+actor {
+
+  let pending_requests = Map.HashMap<Principal, Bool>(10, Principal.equal, Principal.hash);
+
+  private func guard(principal : Principal) : Result.Result<Principal, Error> {
+    if (pending_requests.get(principal) != null) {
+      #err(Error.reject("Already processing a request for principal " # Principal.toText(principal)));
+    } else {
+      pending_requests.put(principal, true);
+      #ok principal;
+    };
+  };
+
+  private func drop(principal : Principal) : () {
+    ignore pending_requests.remove(principal);
+  };
+
+  public shared ({caller}) func example_call_with_locking_per_caller() : async Result.Result<(), (Error.ErrorCode, Text)> {
+    var success = false;
+    try {
+      // Try to create a lock for `caller`, return an error immediately if there is already a call in progress for `caller`
+      switch (guard(caller)) {
+        case (#err e) {
+          return #err(Error.code(e), Error.message(e));
+        };
+        case (#ok _) {
+          success := true;
+        };
+      };
```

Comment:
I was wondering if we could make this shorter. IIUC this could be reduced to: 
```
guard(caller);
success := true; 
```
if we changed `guard` to `throw` instead of returning an error. Then the catch would return the error. But honestly I don't like that because it makes the error handling less explicit and relies on `catch`, making the whole thing more error prone IMO. 

Since I don't really see another way of making it shorter, I suggest we leave it as is. Agreed? Do you maybe see a nice way to shorten this @ggreif ?

Reply:
I think it is shorter if written as
``` Motoko
      switch (guard caller) {
        case (#ok _) success := true;
        case e e;
        };
      };
```